File: python/3-fibonacci_index/fibonacci_index.py
import sys


def main():
    print("This application calculates F(n) - Fibonacci value for the given number.")
    index = process_input()

    print("F({}) = {}".format(index, calculate_fibonacci(index)))


# This function reads users input either from command line arguments or as line input.
def process_input():
    if len(sys.argv) > 1:
        index = sys.argv[1]
    else:
        index = input("Please provide a value n (>= 0): ")

    if not index.isdigit() or int(index) < 0:
        print("Error: The given value must be either a zero or a positive integer.")
        sys.exit(1)

    return int(index)


# By not using recursion it is possible to calculate far larger Fibonacci sequences
# very quickly, e.g. F(10000)
def calculate_fibonacci(index):
    val1 = 0
    val2 = 1
    fib = 0

    for x in range(0, index-1):
        if index == 1:
            return index
        else:
            fib = val1 + val2

            val1 = val2
            val2 = fib

    return fib


# RUn the application
main()

File: golang/3-fibonacci_index/fibonacci_index.go
// Package main contains just one class.
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// re is a regular expression that checks if the input contains only numerical values.
var re = regexp.MustCompile("^[0-9]+")

// main takes user input, validates it and then calls for the Fibonacci calculation.
func main() {
	fmt.Println("This script calculates F(n) - Fibonacci value for the given number.")
	fmt.Println("Please provide a value n (>= 0):")

	// read the user input
	reader := bufio.NewReader(os.Stdin)
	text, _ := reader.ReadString('\n')
	text = strings.Replace(text, "\n", "", -1)

	// use RegExp to validate the input.
	index, err := strconv.ParseUint(text, 10, 16)

	if err != nil {
		fmt.Println("Error: Input conversion failed: ", err)
	}

	// if the input validation is ok, calculate Fibonacci. Otherwise print error message.
	if re.MatchString(text) {
		fmt.Println("F(", index, ") = ", calculateFibonacci(index))
	} else {
		fmt.Println("Error: Please enter either a zero or a positive integer.")
	}
}

// calculateFibonacci uses uint64 to calculate Fibonacci.
// As it is, the largest value for n in F(n) is 65535, which is essentially uint16.
// Might be a bug in func main(), but it still results in pretty large Fibonacci value (it has 18 digits).
func calculateFibonacci(index uint64) uint64 {
	var val1, val2, sum uint64 = 0, 1, 0

	for i := uint64(0); i <= index; i++ {

		if i == 0 || i == 1 {

			if i == index {
				sum = uint64(index)
				break
			}
		} else {
			sum = val1 + val2

			val1 = val2
			val2 = sum
		}
	}

	return sum
}
